File: ui/adminconsole-webapp/net.bluemind.ui.adminconsole.base/src/net/bluemind/ui/adminconsole/base/client/BreadCrumb.java
/* BEGIN LICENSE
 * Copyright © Blue Mind SAS, 2012-2016
 *
 * This file is part of BlueMind. BlueMind is a messaging and collaborative
 * solution.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of either the GNU Affero General Public License as
 * published by the Free Software Foundation (version 3 of the License).
 *
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * See LICENSE.txt
 * END LICENSE
 */
package net.bluemind.ui.adminconsole.base.client;

import java.util.Stack;

import com.google.gwt.core.client.GWT;
import com.google.gwt.event.dom.client.ClickEvent;
import com.google.gwt.event.dom.client.ClickHandler;
import com.google.gwt.event.shared.HandlerRegistration;
import com.google.gwt.i18n.client.Constants;
import com.google.gwt.resources.client.ClientBundle;
import com.google.gwt.resources.client.CssResource;
import com.google.gwt.user.client.ui.Anchor;
import com.google.gwt.user.client.ui.FlowPanel;

import net.bluemind.ui.adminconsole.base.Actions;

public class BreadCrumb extends FlowPanel {

	private Stack<ClickHandler> handlers;
	private Stack<Anchor> anchors;
	private Stack<HandlerRegistration> regs;

	interface BreadCrumbConstants extends Constants {
		String centralAdmin();
	}

	interface BCStyle extends CssResource {

		String bcItem();

	}

	interface BCBundle extends ClientBundle {

		@Source("BreadCrumb.css")
		BCStyle getStyle();
	}

	private static final BreadCrumbConstants bcc;
	private static final BCBundle bcb;
	private static final BCStyle style;

	static {
		bcc = GWT.create(BreadCrumbConstants.class);
		bcb = GWT.create(BCBundle.class);
		style = bcb.getStyle();
		style.ensureInjected();
	}

	public BreadCrumb() {
		handlers = new Stack<ClickHandler>();
		anchors = new Stack<Anchor>();
		regs = new Stack<HandlerRegistration>();

		Anchor root = new Anchor(bcc.centralAdmin());
		root.setHref("#root");
		root.addClickHandler(new ClickHandler() {

			@Override
			public void onClick(ClickEvent event) {
				Actions.get().showWithParams2("root", null);
			}
		});
		root.addStyleName(style.bcItem());
		add(root);
	}

	public void add(Anchor a, ClickHandler cl) {
		if (!handlers.isEmpty()) {
			final ClickHandler toWrap = handlers.peek();
			final Anchor toMakeClickable = anchors.peek();
			ClickHandler wrapper = new ClickHandler() {

				@Override
				public void onClick(ClickEvent event) {
					toWrap.onClick(event);
					// pop items to the left
				}
			};
			toMakeClickable.addClickHandler(wrapper);
		}
		anchors.add(a);
		handlers.add(cl);
		a.addStyleName(style.bcItem());

		add(a);
	}

	public void clearCrumb() {
		for (int i = getWidgetCount() - 1; i > 0; i--) {
			remove(i);
		}
		handlers.clear();
		anchors.clear();
		for (HandlerRegistration hr : regs) {
			hr.removeHandler();
		}
	}

}

File: ui/closures/net.bluemind.commons.ui.closure/src/main/javascript/net/bluemind/ui/Breadcrumb.js
/* BEGIN LICENSE
 * Copyright © Blue Mind SAS, 2012-2016
 *
 * This file is part of BlueMind. BlueMind is a messaging and collaborative
 * solution.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of either the GNU Affero General Public License as
 * published by the Free Software Foundation (version 3 of the License).
 *
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * See LICENSE.txt
 * END LICENSE
 */
goog.provide("net.bluemind.ui.Breadcrumb");

goog.require("goog.array");
goog.require("goog.ui.Container");
goog.require("goog.ui.Container.Orientation");
goog.require("net.bluemind.ui.BreadcrumbItem");
goog.require("net.bluemind.ui.BreadcrumbRenderer");
goog.require("net.bluemind.ui.BreadcrumbSeparator");

/**
 * @constructor
 * 
 * @param {net.bluemind.ui.BreadcrumbRenderer=} opt_renderer
 * @param {goog.dom.DomHelper=} opt_domHelper
 * @extends {goog.ui.Container}
 */
net.bluemind.ui.Breadcrumb = function(opt_renderer, opt_domHelper) {
  var renderer = opt_renderer || net.bluemind.ui.BreadcrumbRenderer.getInstance();
  goog.ui.Container.call(this, goog.ui.Container.Orientation.HORIZONTAL, renderer, opt_domHelper);
  this.setFocusable(false);
  this.setFocusableChildrenAllowed(true);
}
goog.inherits(net.bluemind.ui.Breadcrumb, goog.ui.Container);

/**
 * Path separator
 * 
 * @type {string}
 * @private
 */
net.bluemind.ui.Breadcrumb.prototype.separator_ = '/';

/**
 * Set breadcrumb full path. The path will be exploded with #getSeparator.
 * 
 * @param {*} path
 */
net.bluemind.ui.Breadcrumb.prototype.setPath = function(path) {
  this.reset();
  var elements = path.split(this.separator_);
  var tmp = this.separator_;
  /** @meaning general.userhome */
  var MSG_HOME = goog.getMsg('Home');
  var child = new net.bluemind.ui.BreadcrumbItem(MSG_HOME);
  child.setModel(tmp);
  child.setId(tmp);
  this.addChild(child, true);
  goog.array.forEach(elements, function(element) {
    if (element != '') {
      this.addChild(new net.bluemind.ui.BreadcrumbSeparator(), true);
      tmp += element + this.separator_;
      var child = new net.bluemind.ui.BreadcrumbItem(element);
      child.setModel(tmp);
      child.setId(tmp);
      this.addChild(child, true);
    }
  }, this);
};

/** @override */
net.bluemind.ui.Breadcrumb.prototype.addChildAt = function(control, index, opt_render) {
  goog.base(this, 'addChildAt', control, index, opt_render);
  this.getChildAt(this.getChildCount() - 1).setEnabled(false);
  if (this.getChildCount() > 1) {
    this.getChildAt(this.getChildCount() - 2).setEnabled(true);
  }
}

/** @override */
net.bluemind.ui.Breadcrumb.prototype.removeChild = function(control, opt_render) {
  var child = goog.base(this, 'removeChild', control, opt_render);
  if (this.getChildCount() > 0) {
    this.getChildAt(this.getChildCount() - 1).setEnabled(false);
  }
  return child;
}

/**
 * Reset breadcrumb.
 */
net.bluemind.ui.Breadcrumb.prototype.reset = function() {
  this.forEachChild(function(child) {
    child.dispose();
  });
  this.removeChildren();
};